pkg/controllers/deprovisioning/controller.go

```diff
@@ -0,0 +1,513 @@
+	}
+	oldNode := action.nodesToRemove[0]
+
+	// cordon the node before we launch the replacement to prevent new pods from scheduling to the node
+	if err := c.setNodeUnschedulable(ctx, action.nodesToRemove[0].Name, true); err != nil {
+		return fmt.Errorf("cordoning node %s, %w", oldNode.Name, err)
+	}
+
+	nodeNames, err := c.provisioner.LaunchNodes(ctx, provisioning.LaunchOptions{RecordPodNomination: false}, action.replacementNode)
+	if err != nil {
+		// uncordon the node as the launch may fail (e.g. ICE or incompatible AMI)
+		err = multierr.Append(err, c.setNodeUnschedulable(ctx, oldNode.Name, false))
+		return err
+	}
+	if len(nodeNames) != 1 {
+		// shouldn't ever occur as we are only launching a single node
+		return fmt.Errorf("expected a single node name, got %d", len(nodeNames))
+	}
+
+	metrics.NodesCreatedCounter.WithLabelValues(metrics.DeprovisioningReason).Inc()
+
+	// We have the new node created at the API server so mark the old node for deletion
+	c.cluster.MarkForDeletion(oldNode.Name)
+
+	var k8Node v1.Node
+	// Wait for the node to be ready
+	var once sync.Once
+	if err := retry.Do(func() error {
+		if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: nodeNames[0]}, &k8Node); err != nil {
+			return fmt.Errorf("getting node, %w", err)
+		}
+		once.Do(func() {
+			c.recorder.Publish(LaunchingNode(&k8Node, action.String()))
+		})
+
+		if _, ok := k8Node.Labels[v1alpha5.LabelNodeInitialized]; !ok {
+			// make the user aware of why deprovisioning is paused
+			c.recorder.Publish(WaitingOnReadiness(&k8Node))
+			return fmt.Errorf("node is not initialized")
+		}
+		return nil
+	}, waitRetryOptions...); err != nil {
+		// node never become ready, so uncordon the node we were trying to delete and report the error
+		c.cluster.UnmarkForDeletion(oldNode.Name)
+		return multierr.Combine(c.setNodeUnschedulable(ctx, oldNode.Name, false),
+			fmt.Errorf("timed out checking node readiness, %w", err))
+	}
+	return nil
+}
+
+// calculateLifetimeRemaining calculates the fraction of node lifetime remaining in the range [0.0, 1.0].  If the TTLSecondsUntilExpired
+// is non-zero, we use it to scale down the disruption costs of nodes that are going to expire.  Just after creation, the
+// disruption cost is highest and it approaches zero as the node ages towards its expiration time.
+func (c *Controller) calculateLifetimeRemaining(node candidateNode) float64 {
+	remaining := 1.0
+	if node.provisioner.Spec.TTLSecondsUntilExpired != nil {
+		ageInSeconds := c.clock.Since(node.CreationTimestamp.Time).Seconds()
+		totalLifetimeSeconds := float64(*node.provisioner.Spec.TTLSecondsUntilExpired)
+		lifetimeRemainingSeconds := totalLifetimeSeconds - ageInSeconds
+		remaining = clamp(0.0, lifetimeRemainingSeconds/totalLifetimeSeconds, 1.0)
+	}
+	return remaining
+}
+
+func (c *Controller) setNodeUnschedulable(ctx context.Context, nodeName string, isUnschedulable bool) error {
```

Comment:
helper function?

Reply:
This is a member function still, since it needs the kubeClient. Although I can see it just being passed as a parameter, I plan on keeping it here. WDYT?